pgrx/src/guc.rs

```diff
@@ -114,72 +114,78 @@ where
-    enum_o: Cell<i32>,
-}
-
-impl<T> GucSetting<T> {
-    pub const fn new(value: T) -> Self {
-        GucSetting {
-            value: Cell::new(value),
-            char_p: Cell::new(std::ptr::null_mut()),
-            enum_o: Cell::new(0),
-        }
-    }
+    value: Cell<usize>,
+    boot_val: T,
 }
 
 unsafe impl Sync for GucSetting<bool> {}
 impl GucSetting<bool> {
+    pub const fn new(value: bool) -> Self {
+        GucSetting { value: Cell::new(value as usize), boot_val: value }
+    }
+
     pub fn get(&self) -> bool {
-        self.value.get()
+        self.value.get() == 1
     }
 
-    unsafe fn as_ptr(&self) -> *mut bool {
-        self.value.as_ptr()
+    fn as_ptr(&self) -> *mut bool {
+        self.value.as_ptr() as *mut _
     }
 }
 
 unsafe impl Sync for GucSetting<i32> {}
 impl GucSetting<i32> {
+    pub const fn new(value: i32) -> Self {
+        GucSetting { value: Cell::new(value as usize), boot_val: value }
+    }
+
     pub fn get(&self) -> i32 {
-        self.value.get()
+        self.value.get() as i32
     }
 
-    unsafe fn as_ptr(&self) -> *mut i32 {
-        self.value.as_ptr()
+    fn as_ptr(&self) -> *mut i32 {
+        self.value.as_ptr() as *mut _
     }
 }
 
 unsafe impl Sync for GucSetting<f64> {}
 impl GucSetting<f64> {
+    pub const fn new(value: f64) -> Self {
+        unsafe {
+            GucSetting {
+                value: Cell::new(std::mem::transmute::<f64, u64>(value) as usize),
+                boot_val: value,
+            }
+        }
+    }
+
     pub fn get(&self) -> f64 {
-        self.value.get()
+        f64::from_bits(self.value.get() as u64)
     }
 
-    unsafe fn as_ptr(&self) -> *mut f64 {
-        self.value.as_ptr()
+    fn as_ptr(&self) -> *mut f64 {
+        self.value.as_ptr() as *mut _
     }
 }
 
-unsafe impl Sync for GucSetting<Option<&'static str>> {}
-impl GucSetting<Option<&'static str>> {
-    pub fn get(&self) -> Option<String> {
-        let ptr = self.get_char_ptr();
-        if ptr.is_null() {
-            None
-        } else {
-            let cstr = unsafe { CStr::from_ptr(ptr) };
-            Some(cstr.to_str().unwrap().to_owned())
-        }
+unsafe impl Sync for GucSetting<Option<&'static CStr>> {}
+impl GucSetting<Option<&'static CStr>> {
+    pub const fn new(value: Option<&'static CStr>) -> Self {
+        GucSetting { value: Cell::new(0), boot_val: value }
     }
 
-    pub fn get_char_ptr(&self) -> *mut std::os::raw::c_char {
-        unsafe { *self.char_p.as_ptr() }
+    pub fn get(&self) -> Option<&CStr> {
+        unsafe {
+            if self.value.get() == 0 {
+                None
+            } else {
+                Some(CStr::from_ptr(self.value.get() as *const _))
+            }
+        }
     }
 
-    unsafe fn as_ptr(&self) -> *mut *mut std::os::raw::c_char {
-        self.char_p.as_ptr()
+    fn as_ptr(&self) -> *mut *mut std::os::raw::c_char {
+        self.value.as_ptr() as *mut std::os::raw::c_char as *mut _
```

Comment:
is there a reason `as *mut *mut std::os::raw::c_char` doesn't work?
```suggestion
        self.value.as_ptr() as *mut *mut std::os::raw::c_char
```